pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,247 @@
+	clusterWorkspaceShardInformer tenancyinformers.ClusterWorkspaceShardClusterInformer,
+	apiExportClusterInformer apisinformers.APIExportClusterInformer,
+	kcpClusterClient kcpclientset.ClusterInterface,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue:                         queue,
+		apiExportEndpointSliceLister:  apiExportEndpointSliceInformer.Lister(),
+		apiExportEndpointSliceIndexer: apiExportEndpointSliceInformer.Informer().GetIndexer(),
+		kcpClusterClient:              kcpClusterClient,
+		listClusterWorkspaceShards: func() ([]*tenancyv1alpha1.ClusterWorkspaceShard, error) {
+			return clusterWorkspaceShardInformer.Lister().List(labels.Everything())
+		},
+		getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
+			return apiExportClusterInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		commit: committer.NewCommitter[*APIExportEndpointSlice, Patcher, *APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus](kcpClusterClient.ApisV1alpha1().APIExportEndpointSlices()),
+	}
+
+	apiExportEndpointSliceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIExportEndpointSlice(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExportEndpointSlice(obj)
+		},
+	})
+
+	clusterWorkspaceShardInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueAllAPIExportEndpointSlices(obj)
+			},
+		},
+	)
+
+	return c, nil
+}
+
+type APIExportEndpointSlice = apisv1alpha1.APIExportEndpointSlice
+type APIExportEndpointSliceSpec = apisv1alpha1.APIExportEndpointSliceSpec
+type APIExportEndpointSliceStatus = apisv1alpha1.APIExportEndpointSliceStatus
+type Patcher = apisv1alpha1client.APIExportEndpointSliceInterface
+type Resource = committer.Resource[*APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus]
+type CommitFunc = func(context.Context, *Resource, *Resource) error
+
+// controller reconciles APIExportEndpointSlices. It ensures that the shard endpoints are populated
+//  in the status of every APIExportEndpointSlices.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	apiExportEndpointSliceLister  apislisters.APIExportEndpointSliceClusterLister
+	apiExportEndpointSliceIndexer cache.Indexer
+
+	// kcpClusterClient is cluster aware and used to communicate with kcp API servers
+	kcpClusterClient kcpclientset.ClusterInterface
+
+	listClusterWorkspaceShards func() ([]*tenancyv1alpha1.ClusterWorkspaceShard, error)
```

Comment:
Today all the ClusterWorkspaceShards are in the root workspace of the root shard. I think that the API or its use could be modified so that some information can be surfaced by the shard itself, e.g.:
- resource availability for scheduling
- bound apiexports for filtering out endpoints in APIExportEndpointSlice for shards where no APIBinding has been created for the APIExport
This would  require that ClusterWorkspaceShard or an alternative resource gets created in its own shard, not the root one.

Even if we allow the above I don't think we will filter by workspace in the near future. We would first implement the filtering by ClusterWorkspaceShard labels offered by Partition and PartitionSet.
 
Long story short I think that for now listing across workspaces should be ok. Do you see it differently?